src/xmpp/xmpp-im/xmpp_mammanager.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * xmpp_mammanager.cpp - XEP-0313 Message Archive Management
+ * Copyright (C) 2024 mcneb10
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "xmpp_mammanager.h"
+
+using namespace XMPP;
+
+class MAMManager::Private {
+public:
+    int     mamPageSize;
+    int     mamMaxMessages;
+    bool    flipPages;
+    bool    backwards;
+    Client *client;
+};
+
+MAMManager::MAMManager(Client *client, int mamPageSize, int mamMaxMessages, bool flipPages, bool backwards)
+{
+    d = new Private;
+
+    d->client         = client;
+    d->mamPageSize    = mamPageSize;
+    d->mamMaxMessages = mamMaxMessages;
+    d->flipPages      = flipPages;
+    d->backwards      = backwards;
+}
+
+MAMManager::~MAMManager() { delete d; }
+
+// TODO: review the safety of these methods/object lifetimes
+void MAMManager::getFullArchive(void (*archiveHandler)(QList<QDomElement>), const Jid &j, const bool allowMUCArchives)
+{
+    MAMTask task(d->client->rootTask());
+    connect(&task, &MAMTask::finished, this, [task, archiveHandler]() { archiveHandler(task.archive()); });
+    task.get(j, QString(), QString(), allowMUCArchives, d->mamPageSize, d->mamMaxMessages, d->flipPages, d->backwards);
+}
+
+void MAMManager::getArchiveByIDRange(void (*archiveHandler)(QList<QDomElement>), const Jid &j, const QString &from_id,
```

Comment:
it's  interesting in what case we can use this method.

Reply:
this one can be used to help fetch messages from incomplete archives. for example, when the user wants to scroll up to get more messages or the user reopens the chat and needs to load the newest messages

Reply:
I understand your examples. 
but this method is unlikely to be used because you have more convenient methods below.

maybe will be useful for something. maybe not in Psi.

Reply:
I guess I'll just keep it in there for now, I don't see a reason it needs to be removed